vdr/store/bbolt.go

```diff
@@ -0,0 +1,305 @@
+package store
+
+import (
+	"encoding/json"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/crypto/hash"
+	vdr "github.com/nuts-foundation/nuts-node/vdr/types"
+	"go.etcd.io/bbolt"
+)
+
+type documentVersion struct {
+	Document did.Document
+	Metadata vdr.DocumentMetadata
+}
+
+type documentVersionList struct {
+	Deactivated bool
+	Versions    []hash.SHA256Hash
+}
+
+func (entry documentVersionList) Latest() hash.SHA256Hash {
+	return entry.Versions[len(entry.Versions)-1]
+}
+
+var (
+	documentsBucket = []byte("vdrDocuments")
+	versionsBucket  = []byte("vdrDocumentVersions")
+)
+
+type bboltStore struct {
+	db *bbolt.DB
+}
+
+// NewBBoltStore returns an instance of a BBolt based VDR store
+func NewBBoltStore(db *bbolt.DB) *bboltStore {
+	return &bboltStore{db: db}
+}
+
+func (store *bboltStore) getDocumentKey(id did.DID, hash hash.SHA256Hash) []byte {
+	return append([]byte(id.String()), hash.Slice()...)
+}
+
+func (store *bboltStore) storeDocument(tx *bbolt.Tx, document did.Document, metadata vdr.DocumentMetadata) error {
+	documents, err := tx.CreateBucketIfNotExists(documentsBucket)
+	if err != nil {
+		return err
+	}
+
+	data, err := json.Marshal(documentVersion{
+		Document: document,
+		Metadata: metadata,
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := documents.Put(store.getDocumentKey(document.ID, metadata.Hash), data); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (store *bboltStore) getDocumentVersion(bucket *bbolt.Bucket, id did.DID, hash hash.SHA256Hash) (*documentVersion, error) {
+	data := bucket.Get(store.getDocumentKey(id, hash))
+	if data == nil {
+		return nil, nil
+	}
+
+	result := &documentVersion{}
+
+	if err := json.Unmarshal(data, result); err != nil {
```

Comment:
You could encode this as a byte (for boolean `Deactivated`) + blocks of hashes, since they're always the same length. Then you don't need JSON marshalling. That's how `bboltDAG` does it:

```
// parseHashList splits a list of concatenated hashes into separate hashes.
func parseHashList(input []byte) []hash.SHA256Hash {
	if len(input) == 0 {
		return nil
	}
	num := (len(input) - (len(input) % hash.SHA256HashSize)) / hash.SHA256HashSize
	result := make([]hash.SHA256Hash, num)
	for i := 0; i < num; i++ {
		result[i] = hash.FromSlice(input[i*hash.SHA256HashSize : i*hash.SHA256HashSize+hash.SHA256HashSize])
	}
	return result
}
```

Reply:
I don't like premature optimisations so I ran a quick benchmark:

```
BenchmarkBBoltStore_VersionListBinary-8         32568783                36.84 ns/op
BenchmarkBBoltStore_VersionListJSON-8             383474              3036 ns/op
BenchmarkResolve-8                                105624              9695 ns/op
```

The resolve will drop from `9695 ns/op` to ~ `6696 ns/op` which IMO is significant enough to encode/decode  `documentVersionList` manually.